packages/cli/src/commands/hydrogen/dev-vite.ts

```diff
@@ -0,0 +1,225 @@
+    }),
+    'codegen-config-path': commonFlags.codegenConfigPath,
+    // sourcemap: commonFlags.sourcemap,
+    'disable-virtual-routes': Flags.boolean({
+      description:
+        "Disable rendering fallback routes when a route file doesn't exist.",
+      env: 'SHOPIFY_HYDROGEN_FLAG_DISABLE_VIRTUAL_ROUTES',
+      default: false,
+    }),
+    debug: commonFlags.debug,
+    'inspector-port': commonFlags.inspectorPort,
+    host: Flags.boolean({
+      description: 'Expose the server to the network',
+      default: false,
+      required: false,
+    }),
+    ['env-branch']: commonFlags.envBranch,
+    ['disable-version-check']: Flags.boolean({
+      description: 'Skip the version check when running `hydrogen dev`',
+      default: false,
+      required: false,
+    }),
+    diff: commonFlags.diff,
+  };
+
+  async run(): Promise<void> {
+    const {flags} = await this.parse(DevVite);
+    let directory = flags.path ? path.resolve(flags.path) : process.cwd();
+
+    if (flags.diff) {
+      directory = await prepareDiffDirectory(directory, true);
+    }
+
+    await runDev({
+      ...flagsToCamelObject(flags),
+      path: directory,
+      isLocalDev: flags.diff,
+    });
+  }
+}
+
+type DevOptions = {
+  entry?: string;
+  port: number;
+  path?: string;
+  codegen?: boolean;
+  host?: boolean;
+  codegenConfigPath?: string;
+  disableVirtualRoutes?: boolean;
+  disableVersionCheck?: boolean;
+  envBranch?: string;
+  debug?: boolean;
+  sourcemap?: boolean;
+  inspectorPort: number;
+  isLocalDev?: boolean;
+};
+
+export async function runDev({
+  entry: ssrEntry,
+  port: appPort,
+  path: appPath,
+  host,
+  codegen: useCodegen = false,
+  codegenConfigPath,
+  disableVirtualRoutes,
+  envBranch,
+  debug = false,
+  disableVersionCheck = false,
+  inspectorPort,
+  isLocalDev = false,
+}: DevOptions) {
+  if (!process.env.NODE_ENV) process.env.NODE_ENV = 'development';
+
+  muteDevLogs();
+
+  const root = appPath ?? process.cwd();
+
+  const envPromise = getConfig(root).then(({shop, storefront}) => {
+    const fetchRemote = !!shop && !!storefront?.id;
+    return getAllEnvironmentVariables({root, fetchRemote, envBranch});
+  });
+
+  const vite = await import('vite');
+
+  // Allow Vite to read files from the Hydrogen packages in local development.
+  const fs = isLocalDev
+    ? {allow: [root, fileURLToPath(new URL('../../../../', import.meta.url))]}
+    : undefined;
+
+  const viteServer = await vite.createServer({
+    root,
+    server: {fs, host: host ? true : undefined},
+    ...setH2OPluginContext({
+      cliOptions: {
+        debug,
+        ssrEntry,
+        envPromise,
+        inspectorPort,
+        disableVirtualRoutes,
+      },
+    }),
+  });
+
+  process.once('SIGTERM', async () => {
+    try {
+      await viteServer.close();
+    } finally {
+      process.exit();
+    }
+  });
+
+  if (
+    !viteServer.config.plugins.find((plugin) => plugin.name === 'hydrogen:main')
```

Comment:
Is the hydrogen plugin only necessary for dev mode? There is no error for build when there is no hydrogen plugin.

Reply:
I think so, the plugin is only adding DX and dev tooling. We are also adding config for `ssr.optimizeDeps` but I think it's not used in build.
That said, this is the Hydrogen CLI so maybe we could still enforce it at build time.